### PR TITLE
improve IDE settings change logic on users' behalf from codewhisperer to ensure changes are reverted

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Problem
It seems that sometimes JBPopup object is not disposed right away after it's closed/canceled, therefore the cleanup logic like reverting JB hotkeys overwriten by codewhisperer will not be executed accordingly.

## Solution
1. Dispose the popup manually
2. Use a facade of CodeInsightSettings to manage the changes made by us to make sure changes are reverted

## Regression test

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/1081e208-262c-48cb-91f0-a92b12390da6

Here we're doing option[1]

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/popup/CodeWhispererPopupManager.kt [hunk 1]
```diff
@@ -342,11 +346,6 @@ class CodeWhispererPopupManager {
                 popup.size = popup.preferredContentSize
             }
         } else {
-            val originalAutoPopupCompletionLookup = CodeInsightSettings.getInstance().AUTO_POPUP_COMPLETION_LOOKUP
-            CodeInsightSettings.getInstance().AUTO_POPUP_COMPLETION_LOOKUP = false
-            Disposer.register(popup) {
-                CodeInsightSettings.getInstance().AUTO_POPUP_COMPLETION_LOOKUP = originalAutoPopupCompletionLookup
-            }
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/popup/CodeWhispererPopupManager.kt:349]: duplicate of https://github.com/aws/aws-toolkit-jetbrains/pull/4582/files#diff-4ee308f7dbc11eba350ccc2886bc9cc5a6a3a8590de5b3e7d10a43c18d7bcb27L698

and moved to CodeInsightSettingFacade

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererService.kt [hunk 2]
```diff
@@ -690,16 +692,8 @@ class CodeWhispererService {
...
-        val originalTabExitsBracketsAndQuotes = CodeInsightSettings.getInstance().TAB_EXITS_BRACKETS_AND_QUOTES
-        CodeInsightSettings.getInstance().TAB_EXITS_BRACKETS_AND_QUOTES = false
-        Disposer.register(popup) {
-            CodeInsightSettings.getInstance().TAB_EXITS_BRACKETS_AND_QUOTES = originalTabExitsBracketsAndQuotes
-        }
-        val originalAutoPopupCompletionLookup = CodeInsightSettings.getInstance().AUTO_POPUP_COMPLETION_LOOKUP
-        CodeInsightSettings.getInstance().AUTO_POPUP_COMPLETION_LOOKUP = false
-        Disposer.register(popup) {
-            CodeInsightSettings.getInstance().AUTO_POPUP_COMPLETION_LOOKUP = originalAutoPopupCompletionLookup
-        }
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererService.kt:702]: moved to codeInsightSettingsFacade

**Review by zixlin7 — APPROVED** [review]

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeInsightsSettingsFacade.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        revertAll()
+        val toReverts = mutableListOf<ChangeAndRevert<*>>()
+
+        ChangeAndRevert(settings::TAB_EXITS_BRACKETS_AND_QUOTES, false, parentDisposable).commit().also {
+            toReverts.add(it)
+        }
+
+        ChangeAndRevert(settings::AUTO_POPUP_COMPLETION_LOOKUP, false, parentDisposable).commit().also {
+            toReverts.add(it)
+        }
```
Comment (rli) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeInsightsSettingsFacade.kt:79]: is this section short? is it possible for a user to change these settings while code insight is temporarily disabled?

**Review by rli — APPROVED** [review]

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeInsightsSettingsFacade.kt (hunk 3 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeInsightsSettingsFacade.kt:79]: yea it's short and plus codewhisperer popups will be closed if they're losing focus, so in most cases, popup will be closed and reverted when users opening up the IDE settings panel to configure their CodeInsightSettings. But worth checking out if we have a way to handle that corner case. Will take it as a followup after clearing out other high priority issues.



https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/96e03915-79b7-426d-9837-d4c03b7c8f15

**Pull request merged**